File: javascript/luhn/luhn.js
export const valid = (input) => {
  const cleanedInput = input.replace(/\s+/g, '');
  if (cleanedInput.length <= 1 || cleanedInput.match(/[^\d]/) !== null) {
    return false;
  }

  return [...cleanedInput].reverse()
    .map((c) => Number(c))
    .map((i, idx) => idx % 2 != 0 ? i * 2 : i)
    .map((i) => i > 9 ? i - 9 : i)
    .reduce((acc, i) => acc + i) % 10 == 0;
}

File: ruby/luhn/luhn.rb
module Luhn
  def self.valid?(input)
    input = input.gsub(" ", "")
    return false unless input.match? /^[0-9]{2,}$/
    
    input.reverse.chars.map.with_index do |c, i|
      ci = c.to_i
      i.odd? ? (ci <= 4 ? ci * 2 : ci * 2 - 9) : ci
    end.sum % 10 == 0
  end
end
